Automate updating EOL dates

bwitt opened the pull request:
Automate updating EOL dates by using the https://endoflife.date/docs/api/v1/ API.

* Move EOL dates to `data/eol_dates.json`
* Create script `./bin/update_eol_dates` to update using the API
* Combine centos and centos stream into centos
* Add a workflow `.github/workflows/update-eol-dates.yml` with a cron to create or update a PR via the script using https://github.com/peter-evans/create-pull-request
* use the modulesync bot account
* have tests use the local json file

Fixes https://github.com/voxpupuli/puppet_metadata/issues/190


output

```
$ bundle exec bin/update_eol_dates
Fetching data for AlmaLinux (almalinux)...
  10: 2035-05-31
  9: 2032-05-31
  8: 2029-03-01
Fetching data for Amazon Linux...
  2023: 2029-06-30
  2.0: 2026-06-30
  2018.03: 2023-12-31
  2017.09: 2023-12-31
  2017.03: 2023-12-31
  2016.09: 2023-12-31
  2016.03: 2023-12-31
  2015.09: 2023-12-31
  2015.03: 2023-12-31
  2014.09: 2023-12-31
  2014.03: 2023-12-31
  2013.09: 2023-12-31
  2013.03: 2023-12-31
  2012.09: 2023-12-31
  2012.03: 2023-12-31
  2011.09: 2023-12-31
  2010.11: 2023-12-31
Fetching data for CentOS...
  10: 2030-01-01
  9: 2027-05-31
  8: 2024-05-31
  7: 2024-06-30
  6: 2020-11-30
  5: 2017-03-31
  Preserving 4: 2012-02-29 (not in API)
  Preserving 3: 2010-10-30 (not in API)
Fetching data for Debian (debian)...
  13: 2030-06-30
  12: 2028-06-30
  11: 2026-08-31
  10: 2024-06-30
  9: 2022-07-01
  8: 2020-06-30
  7: 2018-05-31
  6: 2016-02-29
  5: 2012-02-06
  4: 2010-02-15
  3.1: 2008-03-31
  3.0: 2006-06-30
  2.2: 2003-06-30
  2.1: 2000-10-30
  2.0: 1999-02-15
  1.3: 1998-12-08
  1.2: 1997-10-23
  1.1: 1996-12-12
Fetching data for OracleLinux (oracle-linux)...
  10: not yet EOL
  9: 2032-06-30
  8: 2029-07-31
  7: 2024-12-31
  6: 2021-03-31
Fetching data for Fedora (fedora)...
  43: 2026-12-09
  42: 2026-05-13
  41: 2025-12-15
  40: 2025-05-13
  39: 2024-11-26
  38: 2024-05-21
  37: 2023-12-05
  36: 2023-05-16
  35: 2022-12-13
  34: 2022-06-07
  33: 2021-11-30
  32: 2021-05-25
  31: 2020-11-24
  30: 2020-05-26
  29: 2019-11-26
  28: 2019-05-28
  27: 2018-11-30
  26: 2018-05-29
  25: 2017-12-12
  24: 2017-08-08
  23: 2016-12-20
  22: 2016-07-19
  21: 2015-12-01
  20: 2015-06-23
  19: 2015-01-06
  18: 2014-01-14
  17: 2013-07-30
  16: 2013-02-12
  15: 2012-06-26
  14: 2011-12-09
  13: 2011-06-24
  12: 2010-12-02
  11: 2010-06-25
  10: 2009-12-17
  9: 2009-07-10
  8: 2009-01-07
  7: 2008-06-13
  6: 2007-12-07
  5: 2007-07-02
  4: 2006-08-07
  3: 2006-01-16
  2: 2005-04-11
  1: 2004-09-20
Fetching data for FreeBSD (freebsd)...
  15: 2029-12-31
  15.0: 2026-09-30
  14.3: 2026-06-30
  13.5: 2026-04-30
  14.2: 2025-09-30
  13.4: 2025-06-30
  14.1: 2025-03-31
  13.3: 2024-12-31
  14.0: 2024-09-30
  14: 2028-11-30
  13.2: 2024-06-30
  12.4: 2023-12-31
  13.1: 2023-07-31
  12.3: 2023-03-31
  13.0: 2022-08-31
  13: 2026-04-30
  12.2: 2022-03-31
  11.4: 2021-09-30
  12.1: 2021-01-31
  11.3: 2020-09-30
  12.0: 2020-02-29
  12: 2023-12-31
  11.2: 2019-10-31
  10.4: 2018-10-31
  11.1: 2018-09-30
  11.0: 2017-11-30
  11: 2021-09-30
  10.3: 2018-04-30
  10.2: 2016-12-31
  10.1: 2016-12-31
  9.3: 2016-12-31
  10.0: 2015-02-28
  10: 2018-10-31
  9.2: 2014-12-31
  8.4: 2015-08-01
  9.1: 2014-12-31
  8.3: 2014-04-30
  9.0: 2013-03-31
  9: 2016-12-31
  7.4: 2013-02-28
  8.2: 2012-07-31
  8.1: 2012-07-31
  7.3: 2012-03-31
  8.0: 2010-11-30
  8: 2015-08-01
  7.2: 2010-06-30
  7.1: 2011-02-28
  6.4: 2010-11-30
  7.0: 2009-04-30
  7: 2013-02-28
  6.3: 2010-01-31
  6.2: 2008-05-31
  5.5: 2008-05-31
  6.1: 2008-05-31
  6.0: 2007-01-31
  6: 2010-11-30
  5.4: 2006-10-31
  4.11: 2007-01-31
  5.3: 2006-10-31
  4.10: 2007-01-31
  5.2: 2005-02-28
  4.9: 2004-08-31
  5.1: 2004-04-30
  4.8: 2004-02-28
  5.0: 2003-09-30
  5: 2008-05-31
  4.7: 2003-07-31
  4.6: 2003-01-31
  4.5: 2002-09-30
  4.4: 2002-04-30
  4.3: 2001-12-31
  4.2: 2001-07-31
  4.1: 2001-02-28
  3.5: 2001-06-30
  4.0: 2000-10-31
  4: 2007-01-31
  3.4: 2000-09-30
  3.3: 2000-02-29
  3.2: 1999-12-31
  3.1: 1999-08-31
  3.0: 1999-05-31
  2.2: 1998-03-31
  2.1: 1997-06-30
  2.0: 1996-02-28
  1.1: 1995-05-31
  1.0: 1994-08-31
Fetching data for RedHat (rhel)...
  10: 2035-05-31
  9: 2032-05-31
  8: 2029-05-31
  7: 2024-06-30
  6: 2020-11-30
  5: 2017-03-31
  4: 2012-02-29
  Preserving 3: 2010-10-31 (not in API)
Fetching data for Rocky (rocky-linux)...
  10: 2035-05-31
  9: 2032-05-31
  8: 2029-05-31
Preserving Scientific (not in endoflife.date API)
Fetching data for SLES (sles)...
  15.7: 2031-07-31
  15.6: 2025-12-31
  15.5: 2024-12-31
  15.4: 2023-12-31
  15.3: 2022-12-31
  15.2: 2021-12-31
  12.5: 2024-10-31
  15.1: 2021-01-31
  12.4: 2020-06-30
  15.0: 2019-12-31
  12.3: 2019-06-30
  12.2: 2018-03-31
  12.1: 2017-05-31
  11.4: 2019-03-31
  12.0: 2016-06-30
  11.3: 2016-01-31
  11.2: 2014-01-31
  10.4: 2013-07-31
  11.1: 2012-08-31
  10.3: 2011-10-11
  11.0: 2010-12-31
  10.2: 2010-04-11
  10.1: 2008-11-30
  10.0: 2007-12-31
  Preserving 15: 2031-07-31 (not in API)
  Preserving 12: 2024-10-31 (not in API)
  Preserving 11: 2019-03-31 (not in API)
  Preserving 10: 2013-07-31 (not in API)
Fetching data for Ubuntu (LTS only)...
  24.04: 2029-05-31
  22.04: 2027-04-01
  20.04: 2025-05-31
  18.04: 2023-05-31
  16.04: 2021-04-02
  14.04: 2019-04-02
  12.04: 2017-04-28
  10.04: 2013-05-09
  8.04: 2013-05-09
  6.06: 2011-06-01
```

bastelfreak commented on .github/workflows/update-eol-dates.yml:
```diff
@@ -0,0 +1,156 @@
+      - name: Create Pull Request
+        if: steps.update.outputs.changed == 'true' && steps.check-pr.outputs.same != 'true'
+        id: create-pr
+        uses: peter-evans/create-pull-request@v7
```
This action automatically checks for existing PRs and can update them. I don't think you need to implement above logic to check for existing PRs and close them.

bastelfreak commented:
Can you please add `update_eol_dates` to `s.executables` in puppet_metadata.gemspec?

bastelfreak commented:
```
$ bundle exec bin/update_eol_dates
Fetching data for AlmaLinux (almalinux)...
  10: 2035-05-31
  9: 2032-05-31
  8: 2029-03-01
Fetching data for Amazon Linux...
  AL1 2023: 2029-06-30
  AL1 2: 2026-06-30
  AL1 2018.03: 2023-12-31
  AL1 2017.09: 2023-12-31
  AL1 2017.03: 2023-12-31
  AL1 2016.09: 2023-12-31
  AL1 2016.03: 2023-12-31
  AL1 2015.09: 2023-12-31
  AL1 2015.03: 2023-12-31
  AL1 2014.09: 2023-12-31
  AL1 2014.03: 2023-12-31
  AL1 2013.09: 2023-12-31
  AL1 2013.03: 2023-12-31
  AL1 2012.09: 2023-12-31
  AL1 2012.03: 2023-12-31
  AL1 2011.09: 2023-12-31
  AL1 2010.11: 2023-12-31
Failed to fetch data for amazon-linux-2: 404 Not Found
Failed to fetch data for amazon-linux-2023: 404 Not Found
  Preserving 2.0: 2026-06-30 (not in API)
Fetching data for CentOS (centos)...
  8: 2021-12-31
  7: 2024-06-30
  6: 2020-11-30
  5: 2017-03-31
  Preserving 10: 2030-01-01 (not in API)
  Preserving 9: 2027-05-31 (not in API)
  Preserving 4: 2012-02-29 (not in API)
  Preserving 3: 2010-10-30 (not in API)
Fetching data for Debian (debian)...
  13: 2028-08-09
  12: 2026-06-10
  11: 2024-08-14
  10: 2022-09-10
  9: 2020-07-18
  8: 2018-06-17
  7: 2016-04-25
  6: 2014-05-31
  5: 2012-02-06
  4: 2010-02-15
  3.1: 2008-03-31
  3.0: 2006-06-30
  2.2: 2003-06-30
  2.1: 2000-09-30
  2.0: 1999-02-15
  1.3: 1998-12-08
  1.2: 1997-10-23
  1.1: 1996-12-12
Fetching data for OracleLinux (oracle-linux)...
  10: not yet EOL
  9: 2032-06-30
  8: 2029-07-31
  7: 2024-12-31
  6: 2021-03-31
Fetching data for Fedora (fedora)...
  43: 2026-12-09
  42: 2026-05-13
  41: 2025-11-26
  40: 2025-05-13
  39: 2024-11-26
  38: 2024-05-21
  37: 2023-12-05
  36: 2023-05-16
  35: 2022-12-13
  34: 2022-06-07
  33: 2021-11-30
  32: 2021-05-25
  31: 2020-11-24
  30: 2020-05-26
  29: 2019-11-26
  28: 2019-05-28
  27: 2018-11-30
  26: 2018-05-29
  25: 2017-12-12
  24: 2017-08-08
  23: 2016-12-20
  22: 2016-07-19
  21: 2015-12-01
  20: 2015-06-23
  19: 2015-01-06
  18: 2014-01-14
  17: 2013-07-30
  16: 2013-02-12
  15: 2012-06-26
  14: 2011-12-09
  13: 2011-06-24
  12: 2010-12-02
  11: 2010-06-25
  10: 2009-12-17
  9: 2009-07-10
  8: 2009-01-07
  7: 2008-06-13
  6: 2007-12-07
  5: 2007-07-02
  4: 2006-08-07
  3: 2006-01-16
  2: 2005-04-11
  1: 2004-09-20
Fetching data for FreeBSD (freebsd)...
  14.3: 2026-06-30
  13.5: 2026-04-30
  14.2: 2025-09-30
  13.4: 2025-06-30
  14.1: 2025-03-31
  13.3: 2024-12-31
  14.0: 2024-09-30
  14: 2028-11-30
  13.2: 2024-06-30
  12.4: 2023-12-31
  13.1: 2023-07-31
  12.3: 2023-03-31
  13.0: 2022-08-31
  13: 2026-04-30
  12.2: 2022-03-31
  11.4: 2021-09-30
  12.1: 2021-01-31
  11.3: 2020-09-30
  12.0: 2020-02-29
  12: 2023-12-31
  11.2: 2019-10-31
  10.4: 2018-10-31
  11.1: 2018-09-30
  11.0: 2017-11-30
  11: 2021-09-30
  10.3: 2018-04-30
  10.2: 2016-12-31
  10.1: 2016-12-31
  9.3: 2016-12-31
  10.0: 2015-02-28
  10: 2018-10-31
  9.2: 2014-12-31
  8.4: 2015-08-01
  9.1: 2014-12-31
  8.3: 2014-04-30
  9.0: 2013-03-31
  9: 2016-12-31
  7.4: 2013-02-28
  8.2: 2012-07-31
  8.1: 2012-07-31
  7.3: 2012-03-31
  8.0: 2010-11-30
  8: 2015-08-01
  7.2: 2010-06-30
  7.1: 2011-02-28
  6.4: 2010-11-30
  7.0: 2009-04-30
  7: 2013-02-28
  6.3: 2010-01-31
  6.2: 2008-05-31
  5.5: 2008-05-31
  6.1: 2008-05-31
  6.0: 2007-01-31
  6: 2010-11-30
  5.4: 2006-10-31
  4.11: 2007-01-31
  5.3: 2006-10-31
  4.10: 2007-01-31
  5.2: 2005-02-28
  4.9: 2004-08-31
  5.1: 2004-04-30
  4.8: 2004-02-28
  5.0: 2003-09-30
  5: 2008-05-31
  4.7: 2003-07-31
  4.6: 2003-01-31
  4.5: 2002-09-30
  4.4: 2002-04-30
  4.3: 2001-12-31
  4.2: 2001-07-31
  4.1: 2001-02-28
  3.5: 2001-06-30
  4.0: 2000-10-31
  4: 2007-01-31
  3.4: 2000-09-30
  3.3: 2000-02-29
  3.2: 1999-12-31
  3.1: 1999-08-31
  3.0: 1999-05-31
  2.2: 1998-03-31
  2.1: 1997-06-30
  2.0: 1996-02-28
  1.1: 1995-05-31
  1.0: 1994-08-31
Fetching data for RedHat (rhel)...
  10: 2035-05-31
  9: 2032-05-31
  8: 2029-05-31
  7: 2024-06-30
  6: 2020-11-30
  5: 2017-03-31
  4: 2012-02-29
  Preserving 3: 2010-10-31 (not in API)
Fetching data for Rocky (rocky-linux)...
  10: 2035-05-31
  9: 2032-05-31
  8: 2029-05-31
Preserving Scientific (not in endoflife.date API)
Fetching data for SLES (sles)...
  15.7: 2031-07-31
  15.6: 2025-12-31
  15.5: 2024-12-31
  15.4: 2023-12-31
  15.3: 2022-12-31
  15.2: 2021-12-31
  12.5: 2024-10-31
  15.1: 2021-01-31
  12.4: 2020-06-30
  15.0: 2019-12-31
  12.3: 2019-06-30
  12.2: 2018-03-31
  12.1: 2017-05-31
  11.4: 2019-03-31
  12.0: 2016-06-30
  11.3: 2016-01-31
  11.2: 2014-01-31
  10.4: 2013-07-31
  11.1: 2012-08-31
  10.3: 2011-10-11
  11.0: 2010-12-31
  10.2: 2010-04-11
  10.1: 2008-11-30
  10.0: 2007-12-31
  Preserving 15: 2031-07-31 (not in API)
  Preserving 12: 2024-10-31 (not in API)
  Preserving 11: 2019-03-31 (not in API)
  Preserving 10: 2013-07-31 (not in API)
Fetching data for Ubuntu (ubuntu)...
  25.10: 2026-07-01
  25.04: 2026-01-17
  24.10: 2025-07-10
  24.04: 2029-04-25
  23.10: 2024-07-12
  23.04: 2024-01-20
  22.10: 2023-07-20
  22.04: 2027-04-01
  21.10: 2022-07-14
  21.04: 2022-01-20
  20.10: 2021-07-22
  20.04: 2025-05-31
  19.10: 2020-07-06
  19.04: 2020-01-23
  18.10: 2019-07-18
  18.04: 2023-05-31
  17.10: 2018-07-19
  17.04: 2018-01-13
  16.10: 2017-07-20
  16.04: 2021-04-02
  15.10: 2016-07-28
  15.04: 2016-02-04
  14.10: 2015-07-23
  14.04: 2019-04-02
  13.10: 2014-07-17
  13.04: 2014-01-27
  12.10: 2014-05-16
  12.04: 2017-04-28
  11.10: 2013-05-09
  11.04: 2012-10-28
  10.10: 2012-04-10
  10.04: 2013-05-09
  9.10: 2011-04-30
  9.04: 2010-10-23
  8.10: 2010-04-30
  8.04: 2013-05-09
  7.10: 2009-04-18
  7.04: 2008-10-19
  6.10: 2008-04-26
  6.06: 2011-06-01
  5.10: 2007-04-13
  5.04: 2006-10-31
  4.10: 2006-04-30
```

bastelfreak commented:
is this output supposed to look like this:

```
Fetching data for Amazon Linux...
  AL1 2023: 2029-06-30
  AL1 2: 2026-06-30
```

Should 2 and 2023 be versions for AL1?

bwitt commented:
added to gemspec, fixed amazon linux; also fixed centos (combine centos stream into centos)

```
Fetching data for Amazon Linux...
  2023: 2029-06-30
  2.0: 2026-06-30
  2018.03: 2023-12-31
  2017.09: 2023-12-31
  2017.03: 2023-12-31
  2016.09: 2023-12-31
```

```
Fetching data for CentOS...
  10: 2030-01-01
  9: 2027-05-31
  8: 2024-05-31
  7: 2024-06-30
  6: 2020-11-30
  5: 2017-03-31
  Preserving 4: 2012-02-29 (not in API)
  Preserving 3: 2010-10-30 (not in API)
```

bwitt commented:
Looks like most modules use `2.0` for Amazon Linux 2, so that's what this uses.  (should that be `2`?)

bwitt commented:
also since beaker hostgenerator only generates tests for Ubuntu LTS versions, we should probably only add new Ubuntu LTS versions the metadata

bwitt commented:
same thing with SLES and only adding major versions 11, 12, 15

bwitt commented:
hmm should I split out the Ubuntu and SLES `--add-missing` change?

bastelfreak commented on puppet_metadata.gemspec:
```diff
@@ -13,6 +13,7 @@ Gem::Specification.new do |s|
   s.required_ruby_version = Gem::Requirement.new('>= 3.2', '< 5')
 
   s.executables << 'puppet-metadata'
+  s.executables << 'update_eol_dates'
```
please also update ` s.files`. I think that needs to contain `/data/eol_dates.json` now?

bwitt replied:
done

bastelfreak commented:
> hmm should I split out the Ubuntu and SLES --add-missing change?

That would be great <3

I've some time tomorrow where I can do some final testing. Afterwards we can release it \o/

bwitt commented:
> > hmm should I split out the Ubuntu and SLES --add-missing change?
> 
> That would be great <3

Split into https://github.com/voxpupuli/puppet_metadata/pull/216

bwitt commented on spec/operatingsystem_spec.rb:
```diff
@@ -3,6 +3,43 @@
+  # Mock EOL dates for stable tests
+  let(:mock_eol_dates) do
+    {
+      'CentOS' => {
```
it's now using a subset of the json data

ekohl reviewed (commented):
I really like the concept.

bwitt commented on lib/puppet_metadata/operatingsystem.rb:
```diff
@@ -10,211 +11,13 @@ module PuppetMetadata
-    }.freeze
+    EOL_DATES = JSON.parse(
+      File.read(File.expand_path('../../data/eol_dates.json', __dir__)),
+    ).freeze
```
done

bwitt commented:
rebased, added lazy loading of json data, made the tests quieter
